src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1508 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_INTRUSIVELIST_HPP
+#define SHARE_UTILITIES_INTRUSIVELIST_HPP
+
+#include "memory/allStatic.hpp"
+#include "metaprogramming/enableIf.hpp"
+#include "metaprogramming/logical.hpp"
+#include "utilities/align.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/macros.hpp"
+#include <type_traits>
+
+class IntrusiveListEntry;
+class IntrusiveListImpl;
+
+/**
+ * The IntrusiveList class template provides a doubly-linked list in
+ * which the links between elements are embedded directly into objects
+ * contained in the list.  As a result, there are no copies involved
+ * when inserting objects into the list or referencing list objects,
+ * and removing an object from a list need not involve destroying the
+ * object.
+ *
+ * To be used in a IntrusiveList, an object must have a
+ * IntrusiveListEntry member.  An IntrusiveList is associated with the
+ * class of its elements and the entry member.
+ *
+ * An object can be in multiple lists at the same time, so long as
+ * each list uses a different entry member.  That is, the class of the
+ * object must have multiple IntrusiveListEntry members, one for each
+ * list the object is simultaneously an element.
+ *
+ * The API for IntrusiveList is modelled on the containers provided by
+ * the C++ standard library.  In particular, iteration over the
+ * elements is provided via iterator classes.
+ *
+ * IntrusiveLists support polymorphic elements.  Because the objects
+ * in a list are externally managed, rather than being embedded values
+ * in the list, the actual type of such objects may be more specific
+ * than the list's element type.
```

Comment:
Okay, is there a reason that this shouldn't be true? I assume that what you're saying is that we can have:

```c++
struct Super { IntrusiveListEntry entry; };
struct SubA : public Super {};
struct SubB : public Super {};
void foo() {
  IntrusiveList<Super> my_list; // This my_list may contain SubA, SubB, and Super
}
```

And this seems like it should be true for any reasonable intrusive list in C++.

Reply:
Yes, that works.  I mentioned it because I think it's a consequence of
intrusive data structures that might not be obvious to someone familiar
with things like standard containers or our GrowableArray or the like.  (I've
not found a term for distinguishing those from the intrusive kind.)